pkg/querier/querier.go

```diff
@@ -382,6 +382,51 @@ func (mq multiQuerier) Select(ctx context.Context, _ bool, sp *storage.SelectHin
 	return mq.mergeSeriesSets(result)
 }
 
+func clampLabelsMinTime(spanLog *spanlogger.SpanLogger, startMs, endMs, nowMs, maxLabelQueryLengthMs int64) int64 {
+	return clampSeriesMinTime(spanLog, startMs, endMs, nowMs, maxLabelQueryLengthMs)
+}
+
+func clampSeriesMinTime(spanLog *spanlogger.SpanLogger, startMs, endMs, nowMs, maxLabelQueryLengthMs int64) int64 {
+	if maxLabelQueryLengthMs == 0 {
+		// It's unlimited.
+		return startMs
+	}
+	unsetStartTime := startMs == v1.MinTime.UnixMilli()
+	unsetEndTime := endMs == v1.MaxTime.UnixMilli()
+
+	switch {
+	case unsetStartTime && unsetEndTime:
+		// The user asked for "everything", but that's too expensive.
+		// We clamp the start, since the past likely has more data.
+		// Allow querying into the future because that will likely have much less data.
```

Comment:
**question**: does this produces unnecessary workload for the store-gateway or ingester? if so, it could be cheap enough to clamp here and save the extra work

Reply:
not really. The store-gateway only queries existing blocks. The querier takes care of finding the blocks that fit a range. But there it's just a matter of iterating over all blocks and see which fits in the requested window. The complexity is the same in both cases.

The ingester does a very similar thing to the querier with its on-disk blocks and its in-memory head state.


Now, there can be added load **if** there are actually a lot of blocks "in the future" which contain data. By a lot I'd say more than a day's worth of blocks. But I don't believe that that should happen.